From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: yt6801: fix timer API for Linux 6.18

Adapt the YT6801 driver to the timer API changes in Linux 6.18.

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/fuxi-gmac-phy.c
+++ b/fuxi-gmac-phy.c
@@ -351,15 +351,11 @@
 int fxgmac_phy_timer_init(struct fxgmac_pdata *pdata)
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,15,0))
-    init_timer_key(&pdata->expansion.phy_poll_tm, NULL, 0, "fuxi_phy_link_update_timer", NULL);
+    timer_setup(&pdata->expansion.phy_poll_tm, fxgmac_phy_link_poll, 0);
 #else
-    init_timer_key(&pdata->expansion.phy_poll_tm, 0, "fuxi_phy_link_update_timer", NULL);
+    setup_timer(&pdata->expansion.phy_poll_tm, fxgmac_phy_link_poll, (unsigned long)pdata);
 #endif
     pdata->expansion.phy_poll_tm.expires = jiffies + HZ / 2;
-    pdata->expansion.phy_poll_tm.function = (void *)(fxgmac_phy_link_poll);
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(4,15,0))
-    pdata->expansion.phy_poll_tm.data = (unsigned long)pdata;
-#endif
     add_timer(&pdata->expansion.phy_poll_tm);
 
     DPRINTK("fxgmac_phy_timer started, %lx\\n", jiffies);
--- a/fuxi-os.h
+++ b/fuxi-os.h
@@ -67,6 +67,14 @@
 #define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))
 #endif
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,18,0)
+#ifndef from_timer
+#define from_timer(var, callback_timer, timer_fieldname) \\
+    timer_container_of(var, callback_timer, timer_fieldname)
+#endif
+#define del_timer_sync(timer) timer_delete_sync(timer)
+#endif
+
 #ifndef RHEL_MAJOR
